src/wp-includes/html-api/class-wp-html-doctype-info.php

```diff
@@ -0,0 +1,269 @@
+<?php
+/**
+ * HTML API: WP_HTML_Doctype_Info class
+ *
+ * @package WordPress
+ * @subpackage HTML-API
+ * @since 6.7.0
+ */
+
+/**
+ * Core class used by the HTML API processor to represent a DOCTYPE declaration.
+ *
+ * @since 6.7.0
+ */
+class WP_HTML_Doctype_Info {
+	/**
+	 * The name of the DOCTYPE.
+	 *
+	 * @var string|null
+	 */
+	private $name;
+
+	/**
+	 * The public identifier of the DOCTYPE.
+	 *
+	 * @var string|null
+	 */
+	private $public_identifier;
+
+	/**
+	 * The system identifier of the DOCTYPE.
+	 *
+	 * @var string|null
+	 */
+	private $system_identifier;
+
+	/**
+	 * Whether the DOCTYPE token force-quirks flag is set.
+	 *
+	 * @var bool
+	 */
+	private $force_quirks_flag;
+
+	/**
+	 * Constructor.
+	 *
+	 * The arguments to this constructor correspond to the "DOCTYPE token" as defined in the
+	 * HTML specification.
+	 *
+	 * > DOCTYPE tokens have a name, a public identifier, a system identifier,
+	 * > and a force-quirks flag.
+	 *
+	 * @see https://html.spec.whatwg.org/multipage/parsing.html#tokenization
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param string|null $name              The name of the DOCTYPE.
+	 * @param string|null $public_identifier The public identifier of the DOCTYPE.
+	 * @param string|null $system_identifier The system identifier of the DOCTYPE.
+	 * @param bool        $force_quirks_flag Whether the DOCTYPE token force-quirks flag is set.
+	 */
+	public function __construct(
+		?string $name,
+		?string $public_identifier,
+		?string $system_identifier,
+		bool $force_quirks_flag
+	) {
```

Comment:
I don't love this constructor even though it's probably fine with `@access private`

Did you consider moving the parsing code inside of it? We could replace this with a static creator method like `createFromString( $html, $at, $length )` and parse evereything within this class.

Reply:
moving the parse of the token into this class could help resolve the `force-quirks-flag` information visibility problem as well.

when constructing, as soon as `force-quirks-flag` is activated, we can stop parsing and set `$indicates_document_mode = 'quirks';` and so the four properties can be fully: name, public identifier, system identifier, indicated document mode.

Reply:
I moved the parsing into the constructor, made it access-private, and added some notes that it should really be constructed from a doctype token's `get_modifiable_text`.

The parser now expects what's "in" the doc type, and assumes that whatever it is passed is correct.

These classes become quite closely coupled. That may OK at this time given the narrow usefulness of this class (it's practically internal) and the documentation.

Curious to hear your thoughts.

Reply:
instead of tightly coupling these and recreating some of the pains that Core already has, such as the `$attr attribute list from HTML` in `wp_kses_hair()`, I think it would be worth it to double-parse the DOCTYPE tag in its entirety. We can pass the full token into the class and reparse, as parsing the DOCTYPE declaration is much easier than parsing a tag name, and should only happen at most once during the parse of a document, even if multiple tokens appear (since we can ignore the successive tokens).

Reply:
I _almost_ passed the full token and then changed direction after looking at `get_modifiable_text`.

- get_modifiable_text handles newline normalization and null bytes. This isn't a big deal to duplicate.
- If there were pending updates from `set_modifiable_text`, we'd miss them because we'd be getting the "raw" token.

Thinking about that more, I think it's OK to miss updates because I don't think there should be any updates. Changing the doctype seems very dangerous because it has the potential to alter the document's compat mode which could affect how it's parsed.

I do have concerns about the failure mode. What happens if we don't get a valid doctype string? We can return a quirks mode doctype instance with all empty strings, but it would be nice to indicate somehow that the provided string wasn't a valid doctype. Would it be OK to throw an exception in that case given this constructor is `@access private` and we only expect to ever receive a valid DOCTYPE string from one of the processor classes?

Reply:
thankfully we don't currently support setting modifiable text for doctype declarations.

Reply:
> I do have concerns about the failure mode

it seems like if we can't parse it then we can't make any judgement about the document mode because it's not a real DOCTYPE declaration token. returning `null` and ignoring it seems fitting.